src/fallbacks.jl

```diff
@@ -0,0 +1,121 @@
+
+function Base.iterate(rows::RowIterator, st=1)
+    st > length(rows) && return nothing
+    return ColumnsRow(rows.columns, st), st + 1
+end
+
+function rows(x::T) where {T}
+    if columnaccess(T)
+        return RowIterator(columns(x))
+    else
+        throw(ArgumentError("no default `Tables.rows` implementation for type: $T"))
+    end
+end
+
+# build columns from rows
+"""
+    Tables.allocatecolumn(::Type{T}, len) => returns a column type (usually AbstractVector) w/ size to hold `len` elements
+    
+    Custom column types can override with an appropriate "scalar" element type that should dispatch to their column allocator.
+"""
+allocatecolumn(T, len) = Vector{T}(undef, len)
+
+@inline function allocatecolumns(::Schema{names, types}, len) where {names, types}
+    if @generated
+        vals = Tuple(:(allocatecolumn($(fieldtype(types, i)), len)) for i = 1:fieldcount(types))
+        return :(NamedTuple{names}(($(vals...),)))
+    else
+        return NamedTuple{names}(Tuple(allocatecolumn(fieldtype(types, i), len) for i = 1:fieldcount(types)))
+    end
+end
+
+haslength(x) = x === Base.HasLength() || x === Base.HasShape{1}()
+
+# add! will push! or setindex! a value depending on if the row-iterator HasLength or not
+@inline add!(val, col::Int, nm::Symbol, ::Union{Base.HasLength, Base.HasShape{1}}, nt, row) = setindex!(nt[col], val, row)
+@inline add!(val, col::Int, nm::Symbol, T, nt, row) = push!(nt[col], val)
+
+@inline function buildcolumns(schema, rowitr::T) where {T}
+    L = Base.IteratorSize(T)
+    len = haslength(L) ? length(rowitr) : 0
+    nt = allocatecolumns(schema, len)
+    for (i, row) in enumerate(rowitr)
+        eachcolumn(add!, schema, row, L, nt, i)
+    end
+    return nt
+end
+
+@inline add!(dest::AbstractVector, val, ::Union{Base.HasLength, Base.HasShape{1}}, row) = setindex!(dest, val, row)
+@inline add!(dest::AbstractVector, val, T, row) = push!(dest, val)
+
+@inline function add_or_widen!(dest::AbstractVector{T}, val::S, nm::Symbol, L, row, len, updated) where {T, S}
+    if S === T || val isa T
+        add!(dest, val, L, row)
+        return dest
+    else
+        new = allocatecolumn(Base.promote_typejoin(T, S), max(len, length(dest)))
+        row > 1 && copyto!(new, 1, dest, 1, row - 1)
+        add!(new, val, L, row)
+        updated[] = merge(updated[], NamedTuple{(nm,)}((new,)))
+        return new
+    end
+end
+
+@inline function add_or_widen!(val, col, nm, L, columns, row, len, updated)
+    @inbounds add_or_widen!(columns[col], val, nm, L, row, len, updated)
+    return
+end
+
+# when Tables.schema(x) === nothing
+function buildcolumns(::Nothing, rowitr::T) where {T}
+    state = iterate(rowitr)
+    state === nothing && return NamedTuple()
+    row::eltype(rowitr), st = state
+    names = propertynames(row)
+    L = Base.IteratorSize(T)
+    len = haslength(L) ? length(rowitr) : 0
+    sch = Schema(names, nothing)
+    columns = NamedTuple{names}(Tuple(Union{}[] for _ = 1:length(names)))
```

Comment:
@piever @davidanthoff @nalimilan, I've thought of one more thing we should probably discuss here before merging. In this line, I'm defaulting the arrays to `Union{}[]`, which will then widen to actual encountered types while iterating. Reading back over [this monster thread](https://discourse.julialang.org/t/missing-data-and-namedtuple-compatibility/8136/80) however, I remembered that people have this paralyzing fear of this "collect-style" table building losing their original table's missingness. For example, with the current implementation here, we have:
```julia
julia> f = CSV.File(joinpath(dir, "test_basic.csv"))
CSV.File(/Users/jacobquinn/.julia/dev/CSV/test/testfiles/test_basic.csv, rows=3):
Tables.Schema:
 :col1  Union{Missing, Int64}
 :col2  Union{Missing, Int64}
 :col3  Union{Missing, Int64}

julia> f |> columntable
(col1 = Union{Missing, Int64}[1, 4, 7], col2 = Union{Missing, Int64}[2, 5, 8], col3 = Union{Missing, Int64}[3, 6, 9])

julia> Tables.buildcolumns(nothing, f)
(col1 = [1, 4, 7], col2 = [2, 5, 8], col3 = [3, 6, 9])
```
So the collect-strategy drops the missingness, though it's notable that in the "inferred" case (or rather when the source knows its schema, which should be the vast majority of cases), the sink can pass that schema on just fine. Now, I think we have a few different options of things to do here:
* Provide at least a keyword argument, so that people could do, essentially, `f |> x->DataFrame(x; allowmissing=:all)`, (to use the CSV.jl keyword arg `allowmissing`, which can be `:auto`, `:none`, or `:all`). That would essentially control whether we do `Union{}[]` or `Missing[]` columns by default when collecting.
* What should be the default? `Union{}[]`? Or `Missing[]`; I'm inclined to say `Union{}[]`, because I think people over-worry about this case, whereas in practical cases, it will be very much controlled by individual sinks. Sure a DataFrame might lose missingness, but that's a very easy case to handle (i.e. vcat will auto-promote columns for you). For a CSV.write operation, it couldn't care less what the column types are in the first place! Similarly for a database sink, the table will have been pre-defined for columns to allow null values or not, which is completely independent of uninferrable iterators that might come along.

Reply:
Is the problem here that there is no equivalent to ``fieldtype`` for properties in base? Really if there was a ``propertytypes``, this wouldn't be an issue at all, right? You would look at the first row you get, you could tell that the type of a given property is say ``Union{T,Missing}`` and you could allocate the appropriate array.

Maybe just define that function, provide a default fallback that uses ``fieldtype``, and tell folks that if they use a type that adds a ``getproperty``, they also need to add a method for ``propertytypes``?

Or alternatively, catch the special case of a ``NamedTuple`` and use ``fieldtype`` to do the same thing, and lobby that base adds a ``propertytypes`` function?

Reply:
I don't think that would really solve the problem though of sometimes losing missingness; if my row happens to be `NamedTuple` and none of the individual iterated NamedTuples had a missing value for a column, the resulting column wouldn't allow missingness, even though that might be desirable. That's why I think it might be useful for a user to at least pass something like `DataFrame(x; defaultmissing=true)` and the columns would all pop out like `Union{T, Missing}`.

Reply:
Why not? The field types in the ``NamedTuple`` would be ``Union{T,Missing}``, even if there might never be a ``missing`` value, and if in the sink you could query the field type, you can pick the right column type. At least for tables that come from Query.jl that should definitely work, as long as the translator from ``DataValue`` makes sure that if a field in the named tuple is of type ``DataValue``, the field in the named tuple that is returned by the iterator from Tables is of type ``Union{T,Missing}``. I don't see why that wouldn't work.

And in general, I think this problem would *only* show up in a projection case a la Query.jl, right?

Reply:
Indeed, since `NamedTuple` is a standard parametric type (contrary to `Tuple`), its instances can encode the information about whether a value could have been missing in its parameters. Having a `propertytype` function in Base would make sense.

In the meantime, we could define `propertytype` internally without exporting it, have a method for `NamedTuple` (based on `fieldtype`), and require custom row types to define it.

Reply:
I commented more extensively below (in particular I do think there will always be cases where strictification happens), but I agree here that the result of `columns([NamedTuple{(:a, :b), Tuple{Int, Union{Int, Missing}}}((1, 2))])` should allow missing values in the second column and this could be done via `propertytype`